Handlers/TerritoryHierarchyPartHandler.cs

```diff
@@ -0,0 +1,97 @@
+﻿using Nwazet.Commerce.Models;
+using Nwazet.Commerce.Settings;
+using Orchard.ContentManagement;
+using Orchard.ContentManagement.Handlers;
+using Orchard.Data;
+using Orchard.Environment.Extensions;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web.Routing;
+
+namespace Nwazet.Commerce.Handlers {
+    [OrchardFeature("Territories")]
+    public class TerritoryHierarchyPartHandler : ContentHandler {
+
+        private readonly IContentManager _contentManager;
+
+        public TerritoryHierarchyPartHandler(
+            IRepository<TerritoryHierarchyPartRecord> repository,
+            IContentManager contentManager) {
+
+            _contentManager = contentManager;
+
+            Filters.Add(StorageFilter.For(repository));
+
+            // TerritoryHierarchyPart.TerritoryType must be populated
+            OnInitializing<TerritoryHierarchyPart>((ctx, part) => 
+                part.TerritoryType = part.Settings.GetModel<TerritoryHierarchyPartSettings>().TerritoryType);
+            OnLoaded<TerritoryHierarchyPart>((ctx, part) =>
+                part.TerritoryType = string.IsNullOrWhiteSpace(part.TerritoryType) ? 
+                    part.Settings.GetModel<TerritoryHierarchyPartSettings>().TerritoryType : 
+                    part.TerritoryType);
+
+            //Lazyfield setters
+            OnInitializing<TerritoryHierarchyPart>(PropertySetHandlers);
+            //OnInitializing<TerritoryHierarchyPart>(LazyLoadHandlers);
+            OnLoading<TerritoryHierarchyPart>((ctx, part) => LazyLoadHandlers(part));
+            OnVersioning<TerritoryHierarchyPart>((context, part, newVersionPart) => LazyLoadHandlers(newVersionPart));
+
+            //Handle the presence of territories in a hierarchy: may need to run asynchronously
+            OnRemoving<TerritoryHierarchyPart>(RemoveTerritoriesInHierarchy);
+        }
+
+        protected override void GetItemMetadata(GetContentItemMetadataContext context) {
+            var hierarchy = context.ContentItem.As<TerritoryHierarchyPart>();
+
+            if (hierarchy == null)
+                return;
+
+            context.Metadata.EditorRouteValues = new RouteValueDictionary {
```

Comment:
I don't know how.
If I do for example `nameof(Nwazet.Commerce.Controllers.TerritoryHierarchiesAdminController)` I get "TerritoryHierarchiesAdminController" rather than "TerritoryHierarchiesAdmin".

Reply:
Right. Let's just define constants then.

Reply:
done